### PR TITLE
Generate coverage badges locally

### PR DESCRIPTION
**Pull request by segiddins**
I _think_ this should work? Untested, though (done by trying to reverse-engineer the output we got from shields.io).

### CODE REVIEW

File: lib/jazzy/doc_builder.rb [hunk 1]
```diff
@@ -264,17 +264,35 @@ def self.color_for_coverage(coverage)
...
+          <linearGradient id="b" x2="0" y2="100%">
+            <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
+            <stop offset="1" stop-opacity=".1"/>
+          </linearGradient>
+          <clipPath id="a">
+            <rect width="#{width}" height="20" rx="3" fill="#fff"/>
+          </clipPath>
+          <g clip-path="url(#a)">
+            <path fill="#555" d="M0 0h93v20H0z"/>
+            <path fill="##{color_for_coverage(coverage)}" d="M93 0h#{coverage_length / 10 + 10}v20H93z"/>
```
Comment (jpsim) [inline lib/jazzy/doc_builder.rb]: `coverage_length / 10 + 10` isn't exactly right. For example, this produced `0h10v20H93z` whereas shields.io produced `0h35v20H93z`.

Comment (segiddins) [reply lib/jazzy/doc_builder.rb]: For what coverage?

Comment (segiddins) [reply lib/jazzy/doc_builder.rb]: Ah you know what, I think that should be percent_string_length

Comment (jpsim) [reply lib/jazzy/doc_builder.rb]: Alamofire, among others:

```diff
Bacon::Error: File comparison error `docs/badge.svg` for document_alamofire:
--- DIFF -----------------------------------------------------------------------
-<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="128" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><clipPath id="a"><rect width="128" height="20" rx="3" fill="#fff"/></clipPath><g clip-path="url(#a)"><path fill="#555" d="M0 0h93v20H0z"/><path fill="#97CA00" d="M93 0h35v20H93z"/><path fill="url(#b)" d="M0 0h128v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110"><text x="475" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="830">documentation</text><text x="475" y="140" transform="scale(.1)" textLength="830">documentation</text><text x="1095" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="250">86%</text><text x="1095" y="140" transform="scale(.1)" textLength="250">86%</text></g> </svg>
\ No newline at end of file
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="128" height="20">
+  <linearGradient id="b" x2="0" y2="100%">
+    <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
+    <stop offset="1" stop-opacity=".1"/>
+  </linearGradient>
+  <clipPath id="a">
+    <rect width="128" height="20" rx="3" fill="#fff"/>
+  </clipPath>
+  <g clip-path="url(#a)">
+    <path fill="#555" d="M0 0h93v20H0z"/>
+    <path fill="#97CA00" d="M93 0h10v20H93z"/>
+    <path fill="url(#b)" d="M0 0h128v20H0z"/>
+  </g>
+  <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110">
+    <text x="475" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="830">documentation</text>
+    <text x="475" y="140" transform="scale(.1)" textLength="830">documentation</text>
+    <text x="1095" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="250">86%</text>
+    <text x="1095" y="140" transform="scale(.1)" textLength="250">86%</text>
+  </g>
+</svg>
```

**Comment by segiddins** [comment]
Using the onion skin view on the changes in the submodule, it looks like all the svgs are visually identical to the shields.io badges (only difference should be the formatting of the XML)

File: lib/jazzy/doc_builder.rb [hunk 2]
```diff
@@ -244,38 +244,67 @@ def self.document_markdown(source_module, doc_model, path_to_root)
...
+          </linearGradient>
+          <clipPath id="a">
+            <rect width="#{width}" height="20" rx="3" fill="#fff"/>
+          </clipPath>
+          <g clip-path="url(#a)">
+            <path fill="#555" d="M0 0h93v20H0z"/>
+            <path fill="##{color_for_coverage(coverage)}" d="M93 0h#{percent_string_length / 10 + 10}v20H93z"/>
+            <path fill="url(#b)" d="M0 0h#{width}v20H0z"/>
+          </g>
+          <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110">
```
Comment (jpsim) [inline lib/jazzy/doc_builder.rb]: Is `font-size="110"` a typo? The badges from shields.io use `font-size="11"`.

**Comment by jpsim** [comment]
I'll push a few related changes here, I hope you don't mind.

**Comment by jpsim** [comment]
After applying an XML formatter to both the "before" and "after", I get this diff:

```diff
--- a/badge1.svg
+++ b/badge1.svg
@@ -11,10 +11,10 @@
     <path d="M93 0h35v20H93z" fill="#97CA00"/>
     <path d="M0 0h128v20H0z" fill="url(#b)"/>
   </g>
-  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11" text-anchor="middle">
-    <text fill="#010101" fill-opacity=".3" x="46.5" y="15">documentation</text>
-    <text x="46.5" y="14">documentation</text>
-    <text fill="#010101" fill-opacity=".3" x="109.5" y="15">86%</text>
-    <text x="109.5" y="14">86%</text>
+  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110" text-anchor="middle">
+    <text fill="#010101" fill-opacity=".3" textLength="830" transform="scale(.1)" x="475" y="150">documentation</text>
+    <text textLength="830" transform="scale(.1)" x="475" y="140">documentation</text>
+    <text fill="#010101" fill-opacity=".3" textLength="250" transform="scale(.1)" x="1095" y="150">86%</text>
+    <text textLength="250" transform="scale(.1)" x="1095" y="140">86%</text>
   </g>
 </svg>
--- a/badge2.svg
+++ b/badge2.svg
@@ -11,10 +11,10 @@
     <path d="M93 0h35v20H93z" fill="#a4a61d"/>
     <path d="M0 0h128v20H0z" fill="url(#b)"/>
   </g>
-  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11" text-anchor="middle">
-    <text fill="#010101" fill-opacity=".3" x="46.5" y="15">documentation</text>
-    <text x="46.5" y="14">documentation</text>
-    <text fill="#010101" fill-opacity=".3" x="109.5" y="15">70%</text>
-    <text x="109.5" y="14">70%</text>
+  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110" text-anchor="middle">
+    <text fill="#010101" fill-opacity=".3" textLength="830" transform="scale(.1)" x="475" y="150">documentation</text>
+    <text textLength="830" transform="scale(.1)" x="475" y="140">documentation</text>
+    <text fill="#010101" fill-opacity=".3" textLength="250" transform="scale(.1)" x="1095" y="150">70%</text>
+    <text textLength="250" transform="scale(.1)" x="1095" y="140">70%</text>
   </g>
 </svg>
--- a/badge3.svg
+++ b/badge3.svg
@@ -11,10 +11,10 @@
     <path d="M93 0h43v20H93z" fill="#4c1"/>
     <path d="M0 0h136v20H0z" fill="url(#b)"/>
   </g>
-  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11" text-anchor="middle">
-    <text fill="#010101" fill-opacity=".3" x="46.5" y="15">documentation</text>
-    <text x="46.5" y="14">documentation</text>
-    <text fill="#010101" fill-opacity=".3" x="113.5" y="15">100%</text>
-    <text x="113.5" y="14">100%</text>
+  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110" text-anchor="middle">
+    <text fill="#010101" fill-opacity=".3" textLength="830" transform="scale(.1)" x="475" y="150">documentation</text>
+    <text textLength="830" transform="scale(.1)" x="475" y="140">documentation</text>
+    <text fill="#010101" fill-opacity=".3" textLength="330" transform="scale(.1)" x="1135" y="150">100%</text>
+    <text textLength="330" transform="scale(.1)" x="1135" y="140">100%</text>
   </g>
 </svg>
--- a/badge4.svg
+++ b/badge4.svg
@@ -11,10 +11,10 @@
     <path d="M93 0h43v20H93z" fill="#4c1"/>
     <path d="M0 0h136v20H0z" fill="url(#b)"/>
   </g>
-  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11" text-anchor="middle">
-    <text fill="#010101" fill-opacity=".3" x="46.5" y="15">documentation</text>
-    <text x="46.5" y="14">documentation</text>
-    <text fill="#010101" fill-opacity=".3" x="113.5" y="15">100%</text>
-    <text x="113.5" y="14">100%</text>
+  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110" text-anchor="middle">
+    <text fill="#010101" fill-opacity=".3" textLength="830" transform="scale(.1)" x="475" y="150">documentation</text>
+    <text textLength="830" transform="scale(.1)" x="475" y="140">documentation</text>
+    <text fill="#010101" fill-opacity=".3" textLength="330" transform="scale(.1)" x="1135" y="150">100%</text>
+    <text textLength="330" transform="scale(.1)" x="1135" y="140">100%</text>
   </g>
 </svg>
--- a/badge5.svg
+++ b/badge5.svg
@@ -11,10 +11,10 @@
     <path d="M93 0h35v20H93z" fill="#4c1"/>
     <path d="M0 0h128v20H0z" fill="url(#b)"/>
   </g>
-  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11" text-anchor="middle">
-    <text fill="#010101" fill-opacity=".3" x="46.5" y="15">documentation</text>
-    <text x="46.5" y="14">documentation</text>
-    <text fill="#010101" fill-opacity=".3" x="109.5" y="15">96%</text>
-    <text x="109.5" y="14">96%</text>
+  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110" text-anchor="middle">
+    <text fill="#010101" fill-opacity=".3" textLength="830" transform="scale(.1)" x="475" y="150">documentation</text>
+    <text textLength="830" transform="scale(.1)" x="475" y="140">documentation</text>
+    <text fill="#010101" fill-opacity=".3" textLength="250" transform="scale(.1)" x="1095" y="150">96%</text>
+    <text textLength="250" transform="scale(.1)" x="1095" y="140">96%</text>
   </g>
 </svg>
--- a/badge6.svg
+++ b/badge6.svg
@@ -11,10 +11,10 @@
     <path d="M93 0h35v20H93z" fill="#dfb317"/>
     <path d="M0 0h128v20H0z" fill="url(#b)"/>
   </g>
-  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11" text-anchor="middle">
-    <text fill="#010101" fill-opacity=".3" x="46.5" y="15">documentation</text>
-    <text x="46.5" y="14">documentation</text>
-    <text fill="#010101" fill-opacity=".3" x="109.5" y="15">57%</text>
-    <text x="109.5" y="14">57%</text>
+  <g fill="#fff" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110" text-anchor="middle">
+    <text fill="#010101" fill-opacity=".3" textLength="830" transform="scale(.1)" x="475" y="150">documentation</text>
+    <text textLength="830" transform="scale(.1)" x="475" y="140">documentation</text>
+    <text fill="#010101" fill-opacity=".3" textLength="250" transform="scale(.1)" x="1095" y="150">57%</text>
+    <text textLength="250" transform="scale(.1)" x="1095" y="140">57%</text>
   </g>
 </svg>
```

I'm curious if these differences were just missed, or if you have a rationale for diverging.

**Comment by johnfairh** [comment]
The green version (`transform="scale(.1)"` etc.) looks to me to match the changes from shields.io this week & what is in specs/master according to the diffs.

**Comment by jpsim** [comment]
@johnfairh you're right. Since the badges were fine before, I think it's safe to stick with that format.

@segiddins thanks for the PR, I should be able to wrap it up.

**Comment by jpsim** [comment]
Much nicer. Thanks for doing this Sam.

**Comment by segiddins** [comment]
Thanks for wrapping this up!

**Pull request merged**